File: Objective-C/Tests/NotificationTest.m
//
//  NotificationTest.m
//  CouchbaseLite
//
//  Copyright (c) 2017 Couchbase, Inc All rights reserved.
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//  http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
//

#import "CBLTestCase.h"
#import "CBLDatabase+Internal.h"

@interface NotificationTest : CBLTestCase

@end

@implementation NotificationTest

- (void) testDocumentChange {
    // Create doc1 and doc2
    CBLMutableDocument *doc1 = [self createDocument: @"doc1"];
    [doc1 setValue: @"Scott" forKey: @"name"];
    [self saveDocument: doc1 collection: self.defaultCollection];
    
    CBLMutableDocument *doc2 = [self createDocument: @"doc2"];
    [doc2 setValue: @"Daniel" forKey: @"name"];
    [self saveDocument: doc2 collection: self.defaultCollection];
    
    // Expectation:
    XCTestExpectation* x = [self expectationWithDescription: @"document change"];
    NSMutableSet* docs = [NSMutableSet setWithObjects:@"doc1", @"doc2", @"doc3", nil];
    
    // Add change listeners:
    id block = ^void(CBLDocumentChange* change) {
        [docs removeObject:change.documentID];
        if (docs.count == 0)
            [x fulfill];
    };
    
    id listener1 = [self.defaultCollection addDocumentChangeListenerWithID: @"doc1" listener: block];
    id listener2 = [self.defaultCollection addDocumentChangeListenerWithID: @"doc2" listener: block];
    id listener3 = [self.defaultCollection addDocumentChangeListenerWithID: @"doc3" listener: block];
    
    // Update doc1
    [doc1 setValue: @"Scott Tiger" forKey: @"name"];
    [self saveDocument: doc1 collection: self.defaultCollection];
    
    // Delete doc2
    NSError *error;
    Assert([self.defaultCollection deleteDocument: doc2 error: &error], @"Error deleting: %@", error);
    
    // Create doc3
    CBLMutableDocument *doc3 = [self createDocument: @"doc3"];
    [doc3 setValue: @"Jack" forKey: @"name"];
    [self saveDocument: doc3 collection: self.defaultCollection];
    
    [self waitForExpectationsWithTimeout: kExpTimeout handler: NULL];
    
    // Remove listeners:
    [listener1 remove];
    [listener2 remove];
    [listener3 remove];
}

- (void) testAddSameChangeListeners {
    CBLMutableDocument* doc1 = [self createDocument: @"doc1"];
    [doc1 setValue: @"Scott" forKey: @"name"];
    [self saveDocument: doc1 collection: self.defaultCollection];
    
    // Add change listeners:
    XCTestExpectation* x = [self expectationWithDescription: @"document change"];
    __block NSInteger count = 0;
    id block = ^void(CBLDocumentChange* change) {
        count++;
    };
   
    id listener1 = [self.defaultCollection addDocumentChangeListenerWithID: @"doc1" listener: block];
    id listener2 = [self.defaultCollection addDocumentChangeListenerWithID: @"doc1" listener: block];
    id listener3 = [self.defaultCollection addDocumentChangeListenerWithID: @"doc1" listener: block];
    
    // Update doc1:
    [doc1 setValue: @"Scott Tiger" forKey: @"name"];
    [self saveDocument: doc1 collection: self.defaultCollection];
    
    // Let's wait for 0.5 second to make sure that no more changes fired:
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.5 * NSEC_PER_SEC)),
        dispatch_get_main_queue(), ^{
            if (count == 3)
                [x fulfill];
    });
    
    [self waitForExpectationsWithTimeout: kExpTimeout handler: NULL];
    
    // Remove listeners:
    [listener1 remove];
    [listener2 remove];
    [listener3 remove];
}

- (void) testRemoveDocumentChangeListener {
    CBLMutableDocument *doc1 = [self createDocument: @"doc1"];
    [doc1 setValue: @"Scott" forKey: @"name"];
    [self saveDocument: doc1 collection: self.defaultCollection];
    
    // Add change listener:
    XCTestExpectation* x1 = [self expectationWithDescription: @"document change"];
    id block = ^void(CBLDocumentChange* change) {
        [x1 fulfill];
    };
    
    id listener1 = [self.defaultCollection addDocumentChangeListenerWithID: @"doc1" listener: block];
    AssertNotNil(listener1);
    
    // Update doc1:
    [doc1 setValue: @"Scott Tiger" forKey: @"name"];
    [self saveDocument: doc1 collection: self.defaultCollection];
    
    [self waitForExpectationsWithTimeout: kExpTimeout handler: NULL];
    
    // Remove change listener:
    [listener1 remove];
    
    // Update doc1 again:
    [doc1 setValue: @"Scott Tiger" forKey: @"name"];
    [self saveDocument: doc1 collection: self.defaultCollection];
    
    // Let's wait for 0.5 seconds:
    XCTestExpectation *x2 = [self expectationWithDescription: @"No Changes"];
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.5 * NSEC_PER_SEC)),
        dispatch_get_main_queue(), ^{
            [x2 fulfill];
    });
    [self waitForExpectationsWithTimeout: kExpTimeout handler: NULL];
    
    // Remove again:
    [listener1 remove];
}

@end

File: Swift/Tests/NotificationTest.swift
//
//  NotificationTest.swift
//  CouchbaseLite
//
//  Created by Pasin Suriyentrakorn on 5/23/17.
//  Copyright © 2017 Couchbase. All rights reserved.
//

import XCTest
import CouchbaseLiteSwift


class NotificationTest: CBLTestCase {
    func testDatabaseChange() throws {
        let x = self.expectation(description: "change")
        
        let token = db.addChangeListener { (change) in
            XCTAssertEqual(change.documentIDs.count, 10)
            x.fulfill()
        }
       
        try db.inBatch {
            for i in 0...9 {
                let doc = createDocument("doc-\(i)")
                doc.setValue("demo", forKey: "type")
                try saveDocument(doc)
            }
        }
        waitForExpectations(timeout: 5, handler: nil)
        
        db.removeChangeListener(withToken: token)
    }
    
    
    func testDocumentChange() throws {
        var doc1 = createDocument("doc1")
        doc1.setValue("Scott", forKey: "name")
        let savedDoc1 = try saveDocument(doc1)
        
        let doc2 = createDocument("doc2")
        doc2.setValue("Daniel", forKey: "name")
        let savedDoc2 = try saveDocument(doc2)
        
        let x = self.expectation(description: "Got all changes")
        
        var changes = Set<String>()
        changes.insert("doc1")
        changes.insert("doc2")
        changes.insert("doc3")
        
        let handler = { (change: DocumentChange) in
            changes.remove(change.documentID)
            if changes.count == 0 {
                x.fulfill()
            }
        }
        
        // Add change listeners:
        let listener1 = db.addDocumentChangeListener(withID: "doc1", listener: handler)
        let listener2 = db.addDocumentChangeListener(withID: "doc2", listener: handler)
        let listener3 = db.addDocumentChangeListener(withID: "doc3", listener: handler)
        
        // Update doc1:
        doc1 = savedDoc1.toMutable()
        doc1.setValue("Scott Tiger", forKey: "name")
        try saveDocument(doc1)
        
        // Delete doc2:
        try db.deleteDocument(savedDoc2)
        
        // Create doc3:
        let doc3 = createDocument("doc3")
        doc3.setValue("Jack", forKey: "name")
        try saveDocument(doc3)
        
        waitForExpectations(timeout: 5, handler: nil)
        
        db.removeChangeListener(withToken: listener1)
        db.removeChangeListener(withToken: listener2)
        db.removeChangeListener(withToken: listener3)
    }
    
    
    func testAddSameChangeListeners() throws {
        var doc1 = createDocument("doc1")
        doc1.setValue("Scott", forKey: "name")
        doc1 = try saveDocument(doc1).toMutable()
        
        let x = self.expectation(description: "Got all changes")
        
        var count = 0
        let handler = { (change: DocumentChange) in
            count = count + 1
            if count == 3 {
                x.fulfill()
            }
        }
        
        // Add change listeners:
        let listener1 = db.addDocumentChangeListener(withID: "doc1", listener: handler)
        let listener2 = db.addDocumentChangeListener(withID: "doc1", listener: handler)
        let listener3 = db.addDocumentChangeListener(withID: "doc1", listener: handler)
        
        // Update doc1:
        doc1.setValue("Scott Tiger", forKey: "name")
        try saveDocument(doc1)
        
        waitForExpectations(timeout: 5, handler: nil)
        
        db.removeChangeListener(withToken: listener1)
        db.removeChangeListener(withToken: listener2)
        db.removeChangeListener(withToken: listener3)
    }
    
    
    func testRemoveDocumentChangeListener() throws {
        var doc1 = createDocument("doc1")
        doc1.setValue("Scott", forKey: "name")
        doc1 = try saveDocument(doc1).toMutable()
        
        let x1 = self.expectation(description: "change")
        
        // Add change listener:
        
        let token = db.addDocumentChangeListener(withID: "doc1") { (change) in
            x1.fulfill()
        }
        
        // Update doc1:
        doc1.setValue("Scott Tiger", forKey: "name")
        doc1 = try saveDocument(doc1).toMutable()
        
        waitForExpectations(timeout: 5, handler: nil)
        
        // Remove change listener:
        db.removeChangeListener(withToken: token)
        
        doc1.setValue("Scott Tiger", forKey: "name")
        try saveDocument(doc1)
        
        // Let's wait for 0.5 seconds:
        let x2 = expectation(description: "No changes")
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.5) {
            x2.fulfill()
        }
        waitForExpectations(timeout: 5, handler: nil)
        
        // Remove again:
        db.removeChangeListener(withToken: token)
    }
}
